### PR TITLE
Fix up unnecessary async blocking in zome calls

### PR DESCRIPTION
**Pull request by maackle**
### TODO:
- [x] CHANGELOGs updated with appropriate info
- [x] All code changes are reflected in docs, including module-level docs

### CODE REVIEW

File: crates/holochain/src/core/ribosome/guest_callback.rs [hunk 1]
```diff
@@ -4,89 +4,75 @@ pub mod init;
...
 use holochain_types::prelude::*;
+use tokio::task::JoinHandle;
+
+pub type CallStreamItem = Result<(Zome, ExternIO), (Zome, RibosomeError)>;
+pub type CallStream = tokio_stream::wrappers::ReceiverStream<CallStreamItem>;
 
-pub struct CallIterator<R: RibosomeT, I: Invocation> {
+pub fn call_stream<R: RibosomeT + 'static, I: Invocation + 'static>(
     host_context: HostContext,
+    // ribosome: RealRibosome,
```
Comment (maackle) [inline crates/holochain/src/core/ribosome/guest_callback.rs:21]:
```suggestion
```

File: crates/holochain/src/core/ribosome/guest_callback.rs [hunk 2]
```diff
@@ -4,89 +4,75 @@ pub mod init;
...
-    type Item = (Zome, ExternIO);
-    type Error = (Zome, RibosomeError);
-    fn next(&mut self) -> Result<Option<Self::Item>, Self::Error> {
-        Ok(match self.remaining_zomes.first() {
-            Some(zome) => {
-                match self.remaining_components.next() {
-                    Some(to_call) => {
-                        let to_call = to_call.into();
-                        // TODO: some day maybe we can remove this CallIterator fanciness and let this
-                        // be truly async. Or at least make it a CallStream?
```
Comment (maackle) [inline crates/holochain/src/core/ribosome/guest_callback.rs:45]: This PR was inspired by this comment

File: crates/holochain/src/core/ribosome/host_fn/zome_info.rs [hunk 3]
```diff
@@ -16,12 +16,17 @@ pub fn zome_info(
...
             ..
-        } => ribosome
-            .zome_info(call_context.zome.clone())
-            .map_err(|e| match e {
-                RibosomeError::WasmRuntimeError(wasm_error) => wasm_error,
-                other_error => wasm_error!(WasmErrorInner::Host(other_error.to_string())).into(),
-            }),
+        } => {
+            let f = ribosome.zome_info(call_context.zome.clone());
+            tokio_helper::block_on(f, std::time::Duration::from_secs(60))
```
Comment (maackle) [inline crates/holochain/src/core/ribosome/host_fn/zome_info.rs:21]: This is the only place where blocking needed to be reintroduced. This is offset by several places where a blocking thread previously needed to be spawned to accommodate a long-running sync task, and now those tasks are async and can run normally in tokio.

**Review by jost-s — COMMENTED** [review]
Looks good to me. Is there a test for the call iterator?

**Comment by maackle** [comment]
> Looks good to me. Is there a test for the call iterator?

Just the one that was already there: https://github.com/holochain/holochain/pull/4111/files#diff-a7ceb4941355e4712ac6bf2b2ff32ef7fdd6e84069b62b3fbf808445fbf73ed6R75

**Review by neonphog — CHANGES_REQUESTED** [review]
@maackle - I don't know if I read this wrong, but it looks like this has been incorrect even before your change here:

https://github.com/holochain/holochain/blob/a764d8aeb1b5ba8118038db39721f5cfd7860a8e/crates/holochain/src/core/ribosome/real_ribosome.rs#L877-L879

`maybe_call` is being called within a tokio task, and appears to be running a blocking wasm call therefore hanging an entire tokio core thread and ALL tasks that happen to be assigned to that thread until the wasm call completes.

**Comment by maackle** [comment]
@neonphog how does this look? https://github.com/holochain/holochain/pull/4111/commits/7b1a11d680a1af3988c5ceabb0e8ce31af699ba4

**Comment by neonphog** [comment]
@maackle - Yes that looks better. Do you think this has the same problem?:

https://github.com/holochain/holochain/blob/develop/crates/holochain/src/core/ribosome/real_ribosome.rs#L945

**Comment by maackle** [comment]
> Do you think this has the same problem?

@neonphog I guess const_fns are designed to return immediately, so maybe not, but it can't hurt right? So I spawned blocking on that too.

**Review by neonphog — APPROVED** [review]
woot, thanks!

**Pull request merged**